### ISSUE TITLE
Further improvements

### ISSUE THREAD

**Issue by dufferzafar** [lervag/apy#11]
Whenever I start using a new piece of software, I usually have a bunch of design related issues. As in, what would I do differently if I were making it. I'm just opening this issue to be a "catch-all" place for such things.

**Comment by dufferzafar** [comment]
So, I really like the `Editor` based UX. It is very `git commit`-ish.

But, as you know, `git commit` also has the `-m` parameter that allows you to commit with a message inline. Like so: `git commit -m "short and sweet message"`.

I propose that we should have something similar for `apy`. So I can quickly be done from the command line itself. I feel like when I'm adding a note, I want bare minimum friction!

`apy add -d Python -t python -m Basic-22px -f "What arguments does datetime.date take?" -b "datetime.date(year, month, day) - all integers!"`

**Comment by dufferzafar** [comment]
Now, it's odd to expect someone to type all that!

So, We ought to have a way providing defaults - via the config file we already have in place.

Like, we use the default model `Basic` when none is specified. 

Ideally, I want that once I've spent some time setting up my config, the interface should just be:

`apy add -f "this goes on the front" -b "this goes at the back"`

**Comment by lervag** [comment]
> Whenever I start using a new piece of software, I usually have a bunch of design related issues. As in, what would I do differently if I were making it. I'm just opening this issue to be a "catch-all" place for such things. Will keep adding comments as I use apy

If you are opening issues that are aimed at me for improving things, then I would really appreciate to keep them separate. That is, one issue per task/idea/bug. However, if you plan to get involved, then I'll be glad to allow such a "catch-all" issue. I would still propose that you try and keep it structured, to make it easier to keep track of the ideas and the discussion.

> I propose that we should have something similar [to `git commit -m`] for apy. So I can quickly be done from the command line itself. I feel like when I'm adding a note, I want bare minimum friction!

Yes, I agree this is useful!

> So, We ought to have a way providing defaults - via the config file we already have in place.
> ...
> `apy add -f "this goes on the front" -b "this goes at the back"`

Also agreed. On a tangent: I am also planning to add a template feature for adding cards, where the new card template in the editor can be specified.

----

My personal workflow for adding cards: I collect new notes/cards in my personal notebook/wiki. I then regularly add them to Anki by moving them to a new `notes.md` file which I format according to the markdown syntax. This is all very seamless through some personal Vim bindings. I therefore very seldom use the `apy add` command, and instead use the `apy add-from-file`. I also use the `apy review` command often. I study mostly on my phone, where I mark cards that I think needs an update. These are then updated with `apy review`.

**Comment by dufferzafar** [comment]
> If you are opening issues that are aimed at me for improving things, then I would really appreciate to keep them separate. That is, one issue per task/idea/bug. 

Yeah, okay, I can do that. It'll definitely be better for organization as well.

> Also agreed.

I was thinking of creating "aliases", by defining them in the config file:

```json
{
    "base": "~/.local/share/Anki2",

    "aliases": {
        "py": {
            "deck": "Python::Mine",
            "model": "Basic",
            "tags": ["python"]
        },
        "tmux": {
            "deck": "Unix",
            "model": "Basic",
            "tags": ["tmux"]
        }
    }
}
```

Now, `apy -a tmux` should pick up `tmux` alias from the config and use those settings.

Ofc, we would still need to add `-f/--front` and `-b/--back` switches. For other card types we can maybe add `--card type` switch?

Options given at the command line would _merge_ with those present in the config file. So, specifying `apy -a py -t random` will create a card with _both_ `python` and `random` tags.

>  I collect new notes/cards in my personal notebook/wiki. 

I'll look through `apy`'s code once to see how exactly I can create markdown files. Then will see if I too can use this workflow.

**Comment by lervag** [comment]
I'm curious: Should I expect a PR from you on this, or should I consider it a challenge for myself? I'd be very happy to assist you in contributing directly, of course!

**Comment by dufferzafar** [comment]
Sorry for the late reply, this got buried in the emails.

If the approach that I've suggested seems fine to you, then I'd be happy to create a PR. :)

**Comment by lervag** [comment]
I think the approach you suggest seems sensible, so I'd be happy to see a PR. Thanks! :)

**Comment by michaelsjackson** [comment]
Regarding add-from-file are only .md files supported so far?  Can also be images be used in notes?

> My personal workflow for adding cards: I collect new notes/cards in my personal notebook/wiki. I then regularly add them to Anki by moving them to a new `notes.md` file which I format according to the markdown syntax. This is all very seamless through some personal Vim bindings. I therefore very seldom use the `apy add` command, and instead use the `apy add-from-file`. I also use the `apy review` command often. I study mostly on my phone, where I mark cards that I think needs an update. These are then updated with `apy review`.

**Comment by lervag** [comment]
> Regarding add-from-file are only .md files supported so far? Can also images be used in notes?

The extension doesn't matter, but the file format should be markdown. To be more explicit: The format for specifying notes is something like this:

```markdown
model: Basic
tags: marked

# Note 1
## Front
Question?

## Back
Answer.

# Note 2
tag: silly-tag

## Front
Question?

## Back
Answer

# Note 3
model: NewModel
markdown: false (default is true)

## NewFront
FieldOne

## NewBack
FieldTwo

## FieldThree
FieldThree
```

If you have more questions on this, please open a new issue, as this is off topic for the current issue.

Images can be used. In Anki, the fields in a note are always interpreted as html. Thus the text within a `## Front` section should be markdown. Images can then be included with the `<img` tag as expected, and the images should be added to ones `collection.media` folder. Note, if you have `markdown: true` (i.e., the default), then the content of the fields are converted from markdown to html, which means you can use the `![](image.png)` type syntax to add images. As before, the `image.png` file must be manually added to your `collection.media` directory.

**Comment by michaelsjackson** [comment]
I will continue with anki-editor, for now, as this did not work here anyway... until it works, one day. Thanks for your help.

**Comment by michaelsjackson** [comment]
> I will continue with anki-editor, for now, as this did not work here anyway... until it works, one day. Thanks for your help.

No, it works now beautifully, I am using this instead of anki-editor.  Adding here just for completion of the story. Thanks for sharing your work.
